### PR TITLE
Feat: Boosted Creature

### PR DESCRIPTION
**Pull request by marcosvf132**
First of all, I **know** that there is an open PR for this feat, but I've made some changes and add new implementation that was missing on the other PR, and I realized that a new PR was necessary so it can be properly tested.
 The base script I got from the other PR **(Thanks to Nekiro & oualid6496)**, but I did made changes:

- Removed the 'storage' system from the script and added a system via database, with that we can use this systems integrated with the new tibia client.
-  Add what was missing on the other PR: message when killing a boosted monster on the loot call-back and on player login.
-  Add client support to the boosted creature and the features from the old script is still on this new one: The script change the boosted creature only when server changes the day (Server-save). 
- Add migration to the new database changes. (A `schema.sql` need to be done on the future, but can be tested now with migrations already on this PR).
- The script now directly uses the bestiary system to detect monsters and races.
- Add boosted Spawn interval to the boosted creature on the primary and custom world-Spawn.xml

Now the client changes:
- On the xampp/UniServerZ server file so that the client can support the visual boosted creature on the login screen.
>File: www/pages/ws_login.php

On:
``` PHP
	case 'boostedcreature':
		die(json_encode([
			'boostedcreature' => false,
		]));
	break;
	case 'login':
```

Change to:
``` PHP
	case 'boostedcreature':
		$boostDB = $SQL->query('SELECT * FROM ' . $SQL->tableName('boosted_creature'))->fetchAll();
		foreach ($boostDB as $Tableboost) {			
		die(json_encode([
			'boostedcreature' => true,
			'raceid' => intval($Tableboost['raceid'])
		]));
		}
	break;
	case 'login':

```

If there is some bugs or missing features on this system, just comment it.

### CODE REVIEW

**Comment by gabz182** [comment]
Precisa fazer alguma alteração em schema.sql pra testar?

**Comment by marcosvf132** [comment]
> Precisa fazer alguma alteração em schema.sql pra testar?

Só no migration que ta nos commits ja resolve, o schema ainda tem que ser feito.
Don't need schema.sql changes for now, can be tested only with migration changes.

**Comment by gabz182** [comment]
@marcosvf132 
A creatura que aparece no cliente como boosted e a de dentro do game são diferentes.

18:28 Today's boosted creature: Minotaur Archer Boosted creatures yield more experience points, carry more loot than usual and respawn at a faster rate.
18:47 Loot of a minotaur archer: a piece of archer armor, a broken crossbow, 2 piercing bolts, 11 bolts, 33 gold coins (boosted loot)
Loot e XP parece estar funcionando perfeitamente.

https://prnt.sc/ts0lse

**Comment by marcosvf132** [comment]
No log do server aparece qual monstro? Aparece la no comeco. Não sei o que poderia fazer isso acontecer, aqui não acontece. Ja aconteceu uma vez quando rodei duas distro abertas ao mesmo tempo. Se o loot e a mensagem funciona normalmente, tem algo de errado com o client, experimenta reiniciar o client pra ver se acontece esse erro novamente. Talvez o client não atualize o ícone caso seja mudado a criatura no sistema.

**Comment by gabz182** [comment]
> No log do server aparece qual monstro? Aparece la no comeco. Não sei o que poderia fazer isso acontecer, aqui não acontece. Ja aconteceu uma vez quando rodei duas distro abertas ao mesmo tempo. .

Na distro aparece corretamente como ingame.
>> Boosted creature: Minotaur Archer 
Porém quando abro o cliente 12.40 ele mostra outro monstro.

EDIT: O meu cliente ta pegando o boosted creature do Tibia global, kkkk...
Precisa mudar alguma coisa no cliente pra mudar isso?
@marcosogsantos

**Comment by marcosvf132** [comment]
Precisa mudar no arquivo de login do seu servidor, no post eu coloquei onde fica normalmente o arquivo `ws_login.php`

**Comment by gabz182** [comment]
> Precisa mudar no arquivo de login do seu servidor, no post eu coloquei onde fica normalmente o arquivo `ws_login.php`

Descobri oq é, mudei para o endereço do meu site no cliente em "clientWebService=" , agora ta aparecendo certinho.
Nice bro!
Se tiver mais alguma coisa pra testar em relação a esse sistema, é só falar.

File: data/migrations/4.lua [hunk 1]
```diff
@@ -1,3 +1,10 @@
 function onUpdateDatabase()
+	print("> Updating database to version 4 (Boosted creature)")
```
Comment (majestyotbr) [inline data/migrations/4.lua]: print("> Updating database to version 5 (boosted creature)")

Comment (majestyotbr) [reply data/migrations/4.lua]: Missing **data/migrations/5.lua** with:
```
function onUpdateDatabase()
    return false -- true = There are others migrations file | false = this is the last migration file
end
```

Comment (MajestyOTBR) [reply data/migrations/4.lua:2]: I didn't see before that you did `return false -- true = There are others migrations file | false = this is the last migration file` after the migration.
Can you do what i suggested above? Thanks.

**Comment by majestyotbr** [comment]
Entra e participa do discord do projeto:
https://discordapp.com/invite/3NxYnyV

File: data/migrations/4.lua (hunk 1 above)
Comment (marcosvf132) [reply data/migrations/4.lua]: On the `4.lua` just change `false` to `true` and create a `5.lua` file with `return true` too?

Comment (majestyotbr) [reply data/migrations/4.lua]: `4.lua` change to `print("> Updating database to version 5 (boosted creature)")`, change `false` to `true` and create a `5.lua` with
```
function onUpdateDatabase()
    return false -- true = There are others migrations file | false = this is the last migration file
end
```

Comment (majestyotbr) [reply data/migrations/4.lua]: `4.lua` you can remove the comment `-- true = There are others migrations file | false = this is the last migration file` too.

File: data/lib/others/others.lua [hunk 2]
```diff
@@ -3,3 +3,4 @@ dofile('data/lib/others/dawnport_lib.lua')
 dofile('data/lib/others/modal_window_lib.lua')
 dofile('data/lib/others/reward_boss_lib.lua')
 dofile('data/lib/others/bath_tube.lua')
+dofile('data/lib/others/BoostedCreature.lua')
```
Comment (gpedro) [inline data/lib/others/others.lua]: please, use snake_case on file naming

File: data/migrations/4.lua [hunk 3]
```diff
@@ -1,3 +1,10 @@
 function onUpdateDatabase()
-    return false -- true = There are others migrations file | false = this is the last migration file
+	print("> Updating database to version 5 (boosted creature)")
+	db.query([[CREATE TABLE IF NOT EXISTS `boosted_creature` (
+   `boostname` TEXT,   
+   `date` varchar(250)  NOT NULL default '',
```
Comment (gpedro) [inline data/migrations/4.lua:5]: for storing date fields, use timestamp, date or datetime instead.
i don't see a specific reason to keep the history of boosted creatures. maybe we can move this to server_config.

File: data/migrations/4.lua [hunk 4]
```diff
@@ -1,3 +1,10 @@
...
+   `raceid` varchar(250)  NOT NULL default '',
+   PRIMARY KEY  (`date`)
+   ) AS SELECT 42 AS date, "name" AS boostname, 69 AS raceid]])
```
Comment (gpedro) [inline data/migrations/4.lua:8]: what is this?

File: data/events/scripts/monster.lua [hunk 5]
```diff
@@ -15,13 +15,21 @@ function Monster:onDropLoot(corpse)
 		local monsterLoot = mType:getLoot()
 		for i = 1, #monsterLoot do
 			local item = corpse:createLootItem(monsterLoot[i])
+			if (self:getName():lower() == (BoostedCreature.name):lower()) then
```
Comment (gpedro) [inline data/events/scripts/monster.lua]: unnecessary parenthesis.

File: data/events/scripts/monster.lua [hunk 6]
```diff
@@ -15,13 +15,21 @@ function Monster:onDropLoot(corpse)
...
+				local itemBoosted = corpse:createLootItem(monsterLoot[i])
+			end
 			if not item then
 				print('[Warning] DropLoot:', 'Could not add loot item to corpse.')
 			end
 		end
 
 		if player then
-			local text = ("Loot of %s: %s"):format(mType:getNameDescription(), corpse:getContentDescription())
+		local text = {}
```
Comment (gpedro) [inline data/events/scripts/monster.lua]: missing indentation

File: data/events/scripts/monster.lua [hunk 7]
```diff
@@ -15,13 +15,21 @@ function Monster:onDropLoot(corpse)
...
+			if (self:getName():lower() == (BoostedCreature.name):lower()) then
```
Comment (gpedro) [inline data/events/scripts/monster.lua]: unnecessary parenthesis.

File: data/events/scripts/player.lua [hunk 8]
```diff
@@ -760,6 +760,11 @@ function Player:onGainExperience(source, exp, rawExp)
 			self:setStaminaXpBoost(100)
 		end
 	end
+			
+	-- Boosted creature
+	if (source:getName():lower() == BoostedCreature.name:lower()) then
```
Comment (gpedro) [inline data/events/scripts/player.lua]: unnecessary parenthesis.

File: data/lib/others/BoostedCreature.lua [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+end
+
+-- Using bestiary races/monsters.
+dofile("data/modules/scripts/bestiary/bestiary.lua")
+
+function BoostedCreature:startup()
+    	local bodate = db.storeQuery("SELECT `boostname`, `date` FROM `boosted_creature`")
+        local booldnamestamp = result.getStream(bodate, "boostname")
+        local bodatestamp = result.getStream(bodate, "date")
+        result.free(bodate)
```
Comment (gpedro) [inline data/lib/others/BoostedCreature.lua]: wrong indentation

File: data/lib/others/BoostedCreature.lua [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+-- Updated by Marcosvf132
+if not BoostedCreature then
+	BoostedCreature = {name = ""}
+end
+
+-- Using bestiary races/monsters.
+dofile("data/modules/scripts/bestiary/bestiary.lua")
+
+function BoostedCreature:startup()
+    	local bodate = db.storeQuery("SELECT `boostname`, `date` FROM `boosted_creature`")
```
Comment (gpedro) [inline data/lib/others/BoostedCreature.lua]: you need just the latest, doesn't need to fetch all items.

File: data/lib/others/BoostedCreature.lua [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+        local booldnamestamp = result.getStream(bodate, "boostname")
+        local bodatestamp = result.getStream(bodate, "date")
+        result.free(bodate)
+	if (bodatestamp == os.date("%d")) then -- Same day, after restarted server.
```
Comment (gpedro) [inline data/lib/others/BoostedCreature.lua]: unnecessary parenthesis.

File: data/lib/others/BoostedCreature.lua [hunk 12]
```diff
@@ -0,0 +1,31 @@
+-- Base script by Nekiro & oualid6496
+-- Updated by Marcosvf132
+if not BoostedCreature then
+	BoostedCreature = {name = ""}
+end
+
+-- Using bestiary races/monsters.
+dofile("data/modules/scripts/bestiary/bestiary.lua")
```
Comment (gpedro) [inline data/lib/others/BoostedCreature.lua]: you don't need request all bestiary functions, just the assets.lua file

File: data/lib/others/BoostedCreature.lua [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+        self.name = booldnamestamp
+	else -- New day, roll new creature.
+		local cloneTable = {}
+		for name, raceId in pairs(Bestiary.MonstersName) do
+			cloneTable[#cloneTable + 1] = {name, raceId}
+		end
+		local randMonster = nil
+		repeat
+			randMonster = cloneTable[math.random(#cloneTable)]
+		until randMonster[1] ~= booldnamestamp
```
Comment (gpedro) [inline data/lib/others/BoostedCreature.lua]: a better-optimized way to do that is just pick random index from Bestiary.MonstersName table.
like
```lua
local randomIndex = math.random(1, #Bestiary.MonstersName)
local randMonster = Bestiary.MonstersName[randomIndex]
```

File: data/lib/others/BoostedCreature.lua [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+		db.query("UPDATE `boosted_creature` SET `boostname` = '".. randMonster[1] .."', `date` = '".. os.date("%d") .."', `raceid` = '".. randMonster[2] .."'")
```
Comment (gpedro) [inline data/lib/others/BoostedCreature.lua]: are you saving the day of the month? doesn't make any sense...

**Review by gpedro — COMMENTED** [review]
TL;DR:

boosted_creature table should be used inside the server_config. i think we don't need a history data about these creatures.
minor indentation issues and file naming case.

File: data/migrations/4.lua (hunk 3 above)
Comment (marcosvf132) [reply data/migrations/4.lua:5]: I'm using this because on the Spawn.cpp i'm using 'boostname' to detect and grant the monster a boosted respawn.

File: data/events/scripts/monster.lua (hunk 6 above)
Comment (marcosvf132) [reply data/events/scripts/monster.lua]: the 'text' comes just after this line, if monster is boosted or not.

File: data/lib/others/BoostedCreature.lua (hunk 10 above)
Comment (marcosvf132) [reply data/lib/others/BoostedCreature.lua]: It suppose to query 'date' value and 'boostname', both are used on the following lines.

File: data/lib/others/BoostedCreature.lua (hunk 12 above)
Comment (marcosvf132) [reply data/lib/others/BoostedCreature.lua]: I was having some problem calling assets.lua because the global 'bestiary' is only created on bestiary.lua `Bestiary = {}`

File: data/lib/others/BoostedCreature.lua (hunk 14 above)
Comment (marcosvf132) [reply data/lib/others/BoostedCreature.lua]: I'm using this data to prevent a 'new creature roll' if the server restarts by any reason on the day, with this, the boosted creature only change if the server restart on a different day.

File: data/migrations/4.lua (hunk 4 above)
Comment (marcosvf132) [reply data/migrations/4.lua:8]: The script only update the query, it does not create a new table on database, so to update, first we need that the script already have a table inside the DB.

File: data/lib/others/BoostedCreature.lua (hunk 10 above)
Comment (marcosvf132) [reply data/lib/others/BoostedCreature.lua]: Also, it's aways the latest creature, the script only update the DB, it does not create a new table everyday.

**Comment by Geekbuddy** [comment]
Tested and seems to work fine here.

File: data/lib/others/others.lua (hunk 2 above)
Comment (majestyotbr) [reply data/lib/others/others.lua]: dofile('data/lib/others/boosted_creature.lua')

Comment (marcosvf132) [reply data/lib/others/others.lua]: As soon i get home i will fix it!

**Comment by majestyotbr** [comment]
Tested, working. Thanks for your contribution!

**Review by majestyotbr — APPROVED** [review]

**Comment by majestyotbr** [comment]
Updated login.php:
GesiorAAC: https://github.com/opentibiabr/gesioraac-tibia12-login/releases/tag/1.2
MyAAC: https://github.com/opentibiabr/myaac-tibia12-login/releases/tag/1.5
